### ISSUE TITLE
Add predicate (or others) parameters to represent different scheduling algorithms by creating a pod

### ISSUE THREAD

**Issue by wgliang** [kubernetes/kubernetes#59555]
We've always been obsessive about multiple schedulers, but have we ever thought that we could add a parameter when creating a pod, which could be a combination of predicates, either by ignoring a predicate or by specifying a scheduling strategy, etc. Etc., we can still achieve the effect of using different schedulers

**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature

**What happened**:

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by wanghaoran1988** [comment]
We have [scheduler policy config](https://github.com/kubernetes/kubernetes/blob/master/examples/scheduler-policy-config.json) now, what you say is more like multiple policy in a single scheduler.

/cc @kubernetes/sig-scheduling-misc

**Comment by bsalamat** [comment]
This feature request is somewhat interesting, but letting Pods specify what priority and predicates they are subjected to, has its own problems. Just to name a couple:
* It exposes scheduler priority and predicate functions to the core API.
* It could create a chaos in the cluster. For example, a pod may request not to go through anti-affinity predicate and violates other pods anti-affinity rules. A pod may request not to go through resource fit predicate and be scheduled on unexpected nodes. And, many other similar examples which could confuse users and make debugging very hard. This is particularly problematic in multi-tenant clusters.

**Comment by resouer** [comment]
I would also -1 for this request, as it tends to bring huge complexity and uncertainty to the system, and lack of solid use cases. 

A better solution may drop in the picture of new scheduler framework which will enable users to implement robust customized or multiple schedulers by themsleves.

**Comment by wgliang** [comment]
Look forward to the new scheduling framework.
Yes, this is not a good idea. Our SREs are using this function to deal with some urgent scheduling problems. The reason for this issue is that for the current scheduling, almost every user will modify scheduler's code. I think a good scheduling framework should be to try to avoid allowing users to modify its code in depth, but through the plug-in or even directly through the configuration file to achieve in-depth expansion and customization.

**Comment by k82cn** [comment]
Sorry for the delay :)

I think the major requirement is that: different user/tenant/'department' may have different scheduling requirements (e.g. different scheduling configuration). @wgliang , correct me if any mis-understanding.

In our product, we use different way to resolve that: 1. two level scheduling (avoid conflict) + multi-scheduler, 2. scheduling config per Queue (namespace). In k8s, the draft idea is to enhance multi-scheduler to achieve that, by also avoiding the disadvantage in DaemonSet controller.

@bsalamat , does this requirement make sense to you? I'd like to put this into my backlog, will doc the idea/case later :)

**Comment by bsalamat** [comment]
> In k8s, the draft idea is to enhance multi-scheduler to achieve that

@k82cn Yes, that's the idea, but as you can imagine, having multi-scheduler in a cluster is intended for those who really really need it. It is not supported in hosted solutions and probably most users won't try it.

**Comment by wgliang** [comment]
Yes, it is desirable for most users (especially non-developers) to use configuration files or environment variables instead of multi-scheduler.

**Comment by wanghaoran1988** [comment]
A rainstorming idea on this: 
1. We can extend the apiserver(standalone apiserver/ CRD) + contoller that responsible for launch an scheduler for the user. Assume we have a first class resource called "KubeScheduler", if a user need a specific scheduler based on a subset of the priority/predicates funtions(defined by scheduler policy config), the user can create a "KubeScheduler" resource, and the controller will responsible for launch a scheduler for the user.
2. We also need an admission controller that allow we set the schedulerName for the spesfic pods.

@wgliang Any idea?

**Comment by wgliang** [comment]
I understand what you mean. Your suggestion is very interesting, which greatly improves the flexibility of scheduling. In fact, equivalent to the priority/predicates functions of the deterministic list on the scheduling before, which is equivalent to the incoming parameters. But I have two questions of concern:
1. How to control the apiserver to determine a subset of the priority / predicates functions, in the controller to achieve?
2. Will this bring some performance problems?

**Comment by wanghaoran1988** [comment]
> How to control the apiserver to determine a subset of the priority / predicates functions, in the controller to achieve?

@wgliang 
This is still an multiple scheduler solution, instead of create the scheuler by your self, A "KubeScheduler" resource will responsible for launching a scheduler that have an schedule policy configured. the policy contains the subset priority/predicates functions you needed, a admissoin controller is responsible for setting the pod.spec.schedulerName.

**Comment by wgliang** [comment]
@wanghaoran1988 
 OK, I understand what you mean. In other words, in fact, this subset is also determined. However, sometimes this subset is indeterminate to us and may be an arbitrary combination. 
How can we solve it?

**Comment by k82cn** [comment]
> @k82cn Yes, that's the idea, but as you can imagine, having multi-scheduler in a cluster is intended for those who really really need it.

Multi-scheduler is the draft idea in my mind: k8s can launch scheduler based on configuration, similar to timothysc@ 's bootstrap proposal; but I'm also open other options, e.g. use different configuration for different 'Queue' (with a global policy to avoid conflict), or others. let me take sometime to dump it into a doc :)


> However, sometimes this subset is indeterminate to us and may be an arbitrary combination.

@wgliang , any specific user case here :)
